File: repos/rapp-mirror/common/agentcard.ts
/**
 * The agent card — what you learn about an `agent.py` *before* you let it into
 * your brainstem.
 *
 * Sharing agents is the point: forge one, AirDrop it, and it hot-loads into
 * someone else's kernel. But a file that hot-loads is a file that executes, so
 * a received agent is never installed automatically. It is parsed into a card
 * (what it calls itself, what it says it does, what it will run), scanned for
 * capabilities that deserve a human's attention, and held until someone says
 * yes.
 *
 * Pure and dependency-free on purpose: the same logic runs in the main
 * process, in a node:test, and — ported line-for-line — on iOS.
 */

export type Severity = "critical" | "warn";

export interface Finding {
  severity: Severity;
  /** Stable id so a UI can group/explain findings. */
  id: string;
  /** What the agent can do, in the reader's language. */
  detail: string;
  /** 1-based line in the source where it was seen. */
  line: number;
  evidence: string;
}

export type Verdict = "safe" | "review" | "dangerous" | "invalid";

export interface AgentCard {
  ok: boolean;
  verdict: Verdict;
  className: string;
  name: string;
  description: string;
  parameters: { name: string; description: string }[];
  /** The declared procedure, when the sender forged it with the Forge. */
  steps: string[];
  findings: Finding[];
  /** sha256 is not available here without node:crypto; the caller stamps it. */
  lineCount: number;
  /** Why the file could not be read as an agent at all. */
  error?: string;
}

/** Capability probes. Each is a plain regex over source lines — deliberately
 *  conservative: a false positive costs a glance, a false negative costs a
 *  machine. */
const PROBES: { id: string; severity: Severity; re: RegExp; detail: string }[] = [
  {
    id: "exec",
    severity: "critical",
    re: /\b(?:exec|eval)\s*\(/,
    detail: "runs code built at runtime (exec/eval)",
  },
  {
    id: "shell",
    severity: "critical",
    re: /\b(?:subprocess|os\.system|os\.popen|os\.execv|pty\.spawn)\b/,
    detail: "runs shell commands on your machine",
  },
  {
    id: "credentials",
    severity: "critical",
    re: /(?:\.ssh|\.env\b|id_rsa|keychain|\.brainstem_secret|AWS_SECRET|credentials\.json)/i,
    detail: "reaches for credentials or private keys",
  },
  {
    id: "obfuscation",
    severity: "critical",
    re: /base64\s*\.\s*b64decode|codecs\s*\.\s*decode\s*\([^)]*rot13|bytes\.fromhex/,
    detail: "decodes hidden content before running it",
  },
  {
    id: "network",
    severity: "warn",
    re: /\b(?:requests|httpx|urllib|http\.client|socket|aiohttp)\b/,
    detail: "sends or receives data over the network",
  },
  {
    id: "filewrite",
    severity: "warn",
    re: /\b(?:shutil\.rmtree|os\.remove|os\.unlink|\.write_text\s*\(|\.write_bytes\s*\()|open\s*\([^)]*['"][wa]/,
    detail: "writes to or deletes files",
  },
  {
    id: "dynamic-import",
    severity: "warn",
    re: /\b(?:__import__|importlib)\b/,
    detail: "loads other modules dynamically",
  },
  {
    id: "env",
    severity: "warn",
    re: /os\.environ\b/,
    detail: "reads environment variables",
  },
];

/** Comment and docstring lines are noise for a capability scan. */
function isProse(line: string): boolean {
  const t = line.trim();
  return t.startsWith("#") || t === '"""' || t === "'''";
}

const firstMatch = (source: string, re: RegExp): string | undefined =>
  source.match(re)?.[1];

/** Read the metadata dict's `properties` block into a flat parameter list. */
function parseParameters(source: string): { name: string; description: string }[] {
  const block = source.match(/"properties"\s*:\s*\{([\s\S]*?)\n\s*\}\s*,\s*\n\s*"required"/);
  if (!block) return [];
  const out: { name: string; description: string }[] = [];
  const entry = /"([A-Za-z_][\w]*)"\s*:\s*\{[^}]*?"description"\s*:\s*"((?:[^"\\]|\\.)*)"/g;
  for (const m of block[1].matchAll(entry)) {
    out.push({ name: m[1], description: m[2].replace(/\\"/g, '"') });
  }
  return out;
}

/** Pull the forged procedure back out of the rendered agent. */
function parseSteps(source: string): string[] {
  const block = source.match(/steps\s*=\s*\[([\s\S]*?)\]/);
  if (!block) return [];
  return [...block[1].matchAll(/"((?:[^"\\]|\\.)*)"/g)]
    .map((m) => m[1].replace(/\\"/g, '"'))
    .filter(Boolean);
}

/** Inspect a candidate agent file. Never throws — a hostile file is data. */
export function inspectAgentSource(source: string): AgentCard {
  const lines = source.split("\n");
  const base: AgentCard = {
    ok: false,
    verdict: "invalid",
    className: "",
    name: "",
    description: "",
    parameters: [],
    steps: [],
    findings: [],
    lineCount: lines.length,
  };

  if (!source.trim()) return { ...base, error: "the file is empty" };

  const className = firstMatch(source, /class\s+([A-Za-z_]\w*)\s*\(\s*BasicAgent\s*\)/);
  if (!className) {
    return {
      ...base,
      error: "not a RAPP agent: no `class <Name>(BasicAgent)` was found",
    };
  }
  if (!/def\s+perform\s*\(/.test(source)) {
    return { ...base, className, error: "not a RAPP agent: it has no perform() method" };
  }

  const findings: Finding[] = [];
  lines.forEach((line, index) => {
    if (isProse(line)) return;
    for (const probe of PROBES) {
      if (probe.re.test(line)) {
        findings.push({
          severity: probe.severity,
          id: probe.id,
          detail: probe.detail,
          line: index + 1,
          evidence: line.trim().slice(0, 160),
        });
      }
    }
  });

  // One finding per capability is enough to make the decision.
  const seen = new Set<string>();
  const deduped = findings.filter((f) => (seen.has(f.id) ? false : seen.add(f.id)));

  const verdict: Verdict = deduped.some((f) => f.severity === "critical")
    ? "dangerous"
    : deduped.length
      ? "review"
      : "safe";

  return {
    ok: true,
    verdict,
    className,
    name: firstMatch(source, /self\.name\s*=\s*"([^"]+)"/) || className,
    description:
      firstMatch(source, /"description"\s*:\s*"((?:[^"\\]|\\.)*)"/)?.replace(/\\"/g, '"') || "",
    parameters: parseParameters(source),
    steps: parseSteps(source),
    findings: deduped,
    lineCount: lines.length,
  };
}

/** The one-line answer a human reads before deciding. */
export function cardSummary(card: AgentCard): string {
  if (!card.ok) return card.error || "unreadable";
  const what = card.description || "no description";
  switch (card.verdict) {
    case "safe":
      return `${card.className} — ${what}. Runs no shell, no network, no file writes.`;
    case "review":
      return `${card.className} — ${what}. Also: ${card.findings.map((f) => f.detail).join("; ")}.`;
    default:
      return `${card.className} — ${what}. DANGEROUS: ${card.findings
        .filter((f) => f.severity === "critical")
        .map((f) => f.detail)
        .join("; ")}.`;
  }
}

File: repos/rapp-mirror/ios/RappMirrorPackage/Sources/RappMirrorFeature/AgentCard.swift
import Foundation

public enum FindingSeverity: String, Sendable, Codable, Equatable {
    case critical
    case warn
}

public enum AgentVerdict: String, Sendable, Codable, Equatable {
    case safe
    case review
    case dangerous
    case invalid
}

public struct Finding: Sendable, Codable, Equatable, Identifiable {
    public var id: String
    public var severity: FindingSeverity
    public var detail: String
    public var line: Int
    public var evidence: String

    public init(severity: FindingSeverity, id: String, detail: String, line: Int, evidence: String) {
        self.severity = severity
        self.id = id
        self.detail = detail
        self.line = line
        self.evidence = evidence
    }
}

public struct AgentParameter: Sendable, Codable, Equatable, Identifiable {
    public var id: String { name }
    public let name: String
    public let description: String

    public init(name: String, description: String) {
        self.name = name
        self.description = description
    }
}

public struct AgentCard: Sendable, Codable, Equatable, Identifiable {
    public var id: String { className.isEmpty ? name : className }
    public let ok: Bool
    public let verdict: AgentVerdict
    public let className: String
    public let name: String
    public let description: String
    public let parameters: [AgentParameter]
    public let steps: [String]
    public let findings: [Finding]
    public let lineCount: Int
    public let error: String?

    public init(
        ok: Bool,
        verdict: AgentVerdict,
        className: String,
        name: String,
        description: String,
        parameters: [AgentParameter],
        steps: [String],
        findings: [Finding],
        lineCount: Int,
        error: String? = nil
    ) {
        self.ok = ok
        self.verdict = verdict
        self.className = className
        self.name = name
        self.description = description
        self.parameters = parameters
        self.steps = steps
        self.findings = findings
        self.lineCount = lineCount
        self.error = error
    }
}

private struct CapabilityProbe: Sendable {
    let id: String
    let severity: FindingSeverity
    let pattern: String
    let options: NSRegularExpression.Options
    let detail: String
}

public enum AgentCardInspector {
    private static let probes: [CapabilityProbe] = [
        CapabilityProbe(id: "exec", severity: .critical, pattern: #"\b(?:exec|eval)\s*\("#, options: [], detail: "runs code built at runtime (exec/eval)"),
        CapabilityProbe(id: "shell", severity: .critical, pattern: #"\b(?:subprocess|os\.system|os\.popen|os\.execv|pty\.spawn)\b"#, options: [], detail: "runs shell commands on your machine"),
        CapabilityProbe(id: "credentials", severity: .critical, pattern: #"(?:\.ssh|\.env\b|id_rsa|keychain|\.brainstem_secret|AWS_SECRET|credentials\.json)"#, options: [.caseInsensitive], detail: "reaches for credentials or private keys"),
        CapabilityProbe(id: "obfuscation", severity: .critical, pattern: #"base64\s*\.\s*b64decode|codecs\s*\.\s*decode\s*\([^)]*rot13|bytes\.fromhex"#, options: [], detail: "decodes hidden content before running it"),
        CapabilityProbe(id: "network", severity: .warn, pattern: #"\b(?:requests|httpx|urllib|http\.client|socket|aiohttp)\b"#, options: [], detail: "sends or receives data over the network"),
        CapabilityProbe(id: "filewrite", severity: .warn, pattern: #"\b(?:shutil\.rmtree|os\.remove|os\.unlink|\.write_text\s*\(|\.write_bytes\s*\()|open\s*\([^)]*['"][wa]"#, options: [], detail: "writes to or deletes files"),
        CapabilityProbe(id: "dynamic-import", severity: .warn, pattern: #"\b(?:__import__|importlib)\b"#, options: [], detail: "loads other modules dynamically"),
        CapabilityProbe(id: "env", severity: .warn, pattern: #"os\.environ\b"#, options: [], detail: "reads environment variables"),
    ]

    public static func inspectAgentSource(_ source: String) -> AgentCard {
        let lines = source.components(separatedBy: "\n")
        let base = AgentCard(
            ok: false,
            verdict: .invalid,
            className: "",
            name: "",
            description: "",
            parameters: [],
            steps: [],
            findings: [],
            lineCount: lines.count
        )

        guard !source.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else {
            return base.with(error: "the file is empty")
        }

        guard let className = firstMatch(source, #"class\s+([A-Za-z_]\w*)\s*\(\s*BasicAgent\s*\)"#) else {
            return base.with(error: "not a RAPP agent: no `class <Name>(BasicAgent)` was found")
        }

        guard matches(source, #"def\s+perform\s*\("#) else {
            return base.with(className: className, error: "not a RAPP agent: it has no perform() method")
        }

        var findings: [Finding] = []
        var insideDocstring = false
        for (index, line) in lines.enumerated() {
            if isProse(line, insideDocstring: &insideDocstring) { continue }
            for probe in probes where matches(line, probe.pattern, options: probe.options) {
                findings.append(
                    Finding(
                        severity: probe.severity,
                        id: probe.id,
                        detail: probe.detail,
                        line: index + 1,
                        evidence: String(line.trimmingCharacters(in: .whitespacesAndNewlines).prefix(160))
                    )
                )
            }
        }

        var seen = Set<String>()
        let deduped = findings.filter { finding in
            guard !seen.contains(finding.id) else { return false }
            seen.insert(finding.id)
            return true
        }
        let verdict: AgentVerdict = deduped.contains { $0.severity == .critical } ? .dangerous : (deduped.isEmpty ? .safe : .review)
        let name = firstMatch(source, #"self\.name\s*=\s*"([^"]+)""#) ?? className
        let description = firstMatch(source, #""description"\s*:\s*"((?:[^"\\]|\\.)*)""#, options: [.dotMatchesLineSeparators])?.replacingOccurrences(of: #"\""#, with: #"""#) ?? ""

        return AgentCard(
            ok: true,
            verdict: verdict,
            className: className,
            name: name,
            description: description,
            parameters: parseParameters(source),
            steps: parseSteps(source),
            findings: deduped,
            lineCount: lines.count
        )
    }

    public static func cardSummary(_ card: AgentCard) -> String {
        guard card.ok else { return card.error ?? "unreadable" }
        let what = card.description.isEmpty ? "no description" : card.description
        switch card.verdict {
        case .safe:
            return "\(card.className) — \(what). Runs no shell, no network, no file writes."
        case .review:
            return "\(card.className) — \(what). Also: \(card.findings.map(\.detail).joined(separator: "; "))."
        case .dangerous, .invalid:
            let critical = card.findings.filter { $0.severity == .critical }.map(\.detail).joined(separator: "; ")
            return "\(card.className) — \(what). DANGEROUS: \(critical)."
        }
    }

    private static func isProse(_ line: String, insideDocstring: inout Bool) -> Bool {
        let trimmed = line.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.hasPrefix("#") { return true }
        let startsTriple = trimmed.hasPrefix("\"\"\"") || trimmed.hasPrefix("'''")
        guard startsTriple || insideDocstring else { return false }

        if startsTriple {
            let quote = trimmed.hasPrefix("\"\"\"") ? "\"\"\"" : "'''"
            let remainder = String(trimmed.dropFirst(3))
            let closesSameLine = remainder.contains(quote)
            insideDocstring = !insideDocstring && !closesSameLine
        } else if trimmed.contains("\"\"\"") || trimmed.contains("'''") {
            insideDocstring = false
        }
        return true
    }

    private static func parseParameters(_ source: String) -> [AgentParameter] {
        guard let block = firstMatch(source, #""properties"\s*:\s*\{([\s\S]*?)\n\s*\}\s*,\s*\n\s*"required""#, options: [.dotMatchesLineSeparators]) else { return [] }
        let regex = try? NSRegularExpression(pattern: #""([A-Za-z_]\w*)"\s*:\s*\{[^}]*?"description"\s*:\s*"((?:[^"\\]|\\.)*)""#, options: [.dotMatchesLineSeparators])
        let range = NSRange(block.startIndex..<block.endIndex, in: block)
        return regex?.matches(in: block, range: range).compactMap { match in
            guard let nameRange = Range(match.range(at: 1), in: block), let descriptionRange = Range(match.range(at: 2), in: block) else { return nil }
            return AgentParameter(
                name: String(block[nameRange]),
                description: String(block[descriptionRange]).replacingOccurrences(of: #"\""#, with: #"""#)
            )
        } ?? []
    }

    private static func parseSteps(_ source: String) -> [String] {
        guard let block = firstMatch(source, #"steps\s*=\s*\[([\s\S]*?)\]"#, options: [.dotMatchesLineSeparators]) else { return [] }
        let regex = try? NSRegularExpression(pattern: #""((?:[^"\\]|\\.)*)""#)
        let range = NSRange(block.startIndex..<block.endIndex, in: block)
        return regex?.matches(in: block, range: range).compactMap { match in
            guard let stepRange = Range(match.range(at: 1), in: block) else { return nil }
            let step = String(block[stepRange]).replacingOccurrences(of: #"\""#, with: #"""#)
            return step.isEmpty ? nil : step
        } ?? []
    }

    private static func firstMatch(_ source: String, _ pattern: String, options: NSRegularExpression.Options = []) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: options) else { return nil }
        let range = NSRange(source.startIndex..<source.endIndex, in: source)
        guard let match = regex.firstMatch(in: source, range: range), match.numberOfRanges > 1,
              let matchRange = Range(match.range(at: 1), in: source) else { return nil }
        return String(source[matchRange])
    }

    private static func matches(_ source: String, _ pattern: String, options: NSRegularExpression.Options = []) -> Bool {
        guard let regex = try? NSRegularExpression(pattern: pattern, options: options) else { return false }
        return regex.firstMatch(in: source, range: NSRange(source.startIndex..<source.endIndex, in: source)) != nil
    }
}

private extension AgentCard {
    func with(className: String? = nil, error: String) -> AgentCard {
        AgentCard(
            ok: ok,
            verdict: verdict,
            className: className ?? self.className,
            name: name,
            description: description,
            parameters: parameters,
            steps: steps,
            findings: findings,
            lineCount: lineCount,
            error: error
        )
    }
}
